### PR TITLE
AO3-6327 Allow users to block other users from commenting.

### PR DESCRIPTION
**Pull request by tickinginstant**
## Issue

https://otwarchive.atlassian.net/browse/AO3-6327

## Purpose

- Add a `Block` class used to represent one user blocking another.
- Add a `Blocked::UsersController` class used to block users, unblock users, and display all of a user's blocks.
  - Modify the CSS to make the messages displayed in this class look right.
  - Make the `index` action accessible to certain admins.
- Add a validator to check for blocking when a model is saved.
  - Modify the `ActsAsCommentable::CommentMethods` class to skip validation more, so that it's still possible to delete comments or reply to comments when blocking is involved.
- Modify the `CommentsController` to prevent commenting when blocked, or editing an existing comment.
- Add a "Block" button to the user page navigation bar.
- Modify the comment views to:
  - Show a notice when the user can't comment as a result of being blocked.
  - Hide the "Reply" button when a user is blocked.
  - Add a "Block" button when the comment author is not anonymous.
- Add a new method `by_anonymous_creator?`.
  - Replace all existing "Anonymous Creator" checks with `by_anonymous_creator?`.
- Make `default_pseud` an association instead of a method.
- Add a bunch of tests.

### CODE REVIEW

File: config/locales/controllers/en.yml
Comment (sarken) [inline config/locales/controllers/en.yml:17]: Let's say `work's creators` instead of `owners` for these messages.

File: config/locales/controllers/en.yml
Comment (sarken) [inline config/locales/controllers/en.yml:18]: We'd like to go with `Sorry, you have been blocked by that user.` in the places that currently use this phrasing.

File: app/controllers/application_controller.rb
Comment (tickinginstant) [reply app/controllers/application_controller.rb:368]: Yeah. I don't remember the specifics, but there's a cucumber test that involves viewing a work, and then accessing `User.current_user` and changing the preferences. This causes the altered `work_title_format` to be saved as well.

I changed it here rather than calling `reload` in the tests because if there ever ends up being a call to `preference.save` in the `WorksController` (unlikely to happen, but the `WorksController` does have the adult content check), it'd be good to avoid accidentally clobbering the work title format.

File: public/stylesheets/site/2.0/22-system-messages.css
Comment (tickinginstant) [reply public/stylesheets/site/2.0/22-system-messages.css:62]: Are you sure? There are a couple of pre-existing places that set `flash[:notice]` equal to an array, which produces a `ul` inside of an `li`, including:
- the admin skins page
- the co-creator requests page
- the troubleshooting page

This will change the look of those flash messages.

File: app/helpers/comments_helper.rb
Comment (tickinginstant) [reply app/helpers/comments_helper.rb:112]: Do you also want to add checks to the `InboxController#reply` action, in case someone is blocked after they load their inbox but before they click the "Reply" button?

File: public/stylesheets/site/2.0/22-system-messages.css
Comment (sarken) [reply public/stylesheets/site/2.0/22-system-messages.css:62]: Yeah, it's fine. We can give those special treatment for cases where there's only one message to display if we decide we're really bothered by single-item lists, but otherwise actually showing the list formatting makes sense.

I did see two that have no business being an array and should definitely be fixed at some point, but I'm fine with leaving them alone for now.

https://github.com/otwcode/otwarchive/blob/58f92a68bed5a28d9296db631b200be9d51facc3/app/controllers/skins_controller.rb#L117-L122

👆 Which also needs tweaks to the text.

https://github.com/otwcode/otwarchive/blob/58f92a68bed5a28d9296db631b200be9d51facc3/app/controllers/works_controller.rb#L396-L399

👆 Which needs tweaks to become i18n-ready. (I was having a hard time figuring out how to trigger this one; if it's one we see often, we can fix it in this PR rather than live with the oddness.)

Comment (tickinginstant) [reply public/stylesheets/site/2.0/22-system-messages.css]: Okay, I've changed the CSS classes.

FYI, that second one actually isn't an array; it's using `<<` for string concatenation. So its formatting wouldn't be affected by the change.

File: app/helpers/comments_helper.rb
Comment (sarken) [reply app/helpers/comments_helper.rb:112]: Sure, if it's not super fussy!

Comment (tickinginstant) [reply app/helpers/comments_helper.rb:112]: Doing a redirect in that particular case isn't too fussy, I think.

It would be a bit fussier to get it to display an error when javascript is enabled, instead of just failing silently. But since that's the existing behavior of the "Reply" button in comment views, I figured it'd be okay not to handle that case.

(It's a little odd that we're only performing this one check, when there are a bunch of others from the `CommentsController` that should also be performed, like disabled comments and frozen comments and replying to spam. But I think the best way to address that is to move all of those checks into the `CommentPolicy` class and rely on Pundit for both users and admins, which is not really in scope.)

**Review by sarken — COMMENTED** [review]
Just two more nitpicks from me! I do want to find a second reviewer, though, so no rush.

**Review by sarken — APPROVED** [review]

**Review by neuroalien — APPROVED** [review]

File: app/views/blocked/users/index.html.erb
Comment (redsummernight) [inline app/views/blocked/users/index.html.erb:3]: This one also for the other reviewers: will this ever be different from the confirm_block page?

File: features/step_definitions/comment_steps.rb
Comment (redsummernight) [inline features/step_definitions/comment_steps.rb:24]: Can you use `"(.*?)"` in the regex?

File: app/views/blocked/users/index.html.erb
Comment (sarken) [reply app/views/blocked/users/index.html.erb:3]: Yeah, the confirm_block page is eventually going to have options on it for things like "delete comments" or "also mute" or whatever, so the text and formatting will diverge as time goes on.

File: app/controllers/blocked/users_controller.rb
Comment (sarken) [reply app/controllers/blocked/users_controller.rb:19]: That's just because I forgot recs were there when making the images >.>

File: app/validators/not_blocked_validator.rb
Comment (tickinginstant) [reply app/validators/not_blocked_validator.rb:24]: They're currently unused, yes. I added it so that this validator will be usable even in (potential future) cases where the error message is supposed to contain more information than just "That user has blocked you." I can remove them for now if you'd like.

File: features/step_definitions/comment_steps.rb
Comment (tickinginstant) [reply features/step_definitions/comment_steps.rb:24]: I'd really prefer not to, because it would allow this to match to `the work "Hello" by "creator"`. But I can drop the `?` if you think that'd be better.

Comment (redsummernight) [reply features/step_definitions/comment_steps.rb:24]: Yeah, sounds good - since the lazy quantifier is not needed in this case.

**Pull request merged**
